src/qutip_qip/decompose/decompose_general_qubit_gate.py

```diff
@@ -0,0 +1,178 @@
+        for i in reversed(range(len(index_list))):
+            U_non_trivial = np.full([2, 2], None, dtype=complex)
+            index_info = []
+            U_index_together = []
+
+            # create index list
+            index_1, index_2 = index_list[i]
+            index_info = [index_1, index_2]
+            U_index_together.append(index_info)
+
+            # create 2 x 2 arrays
+            U_two_level = array_list[i]
+            U_non_trivial[0][0] = U_two_level[index_1][index_1]
+            U_non_trivial[1][0] = U_two_level[index_2][index_1]
+            U_non_trivial[0][1] = U_two_level[index_1][index_2]
+            U_non_trivial[1][1] = U_two_level[index_2][index_2]
+            U_index_together.append(Qobj(U_non_trivial, dims=[[2] * 1] * 2))
+
+            compact_U_information.append(U_index_together)
+
+        return compact_U_information
+
+
+def _create_dict_for_two_level_arrays(two_level_output):
+    """Creates a dictionary with keys for the total number of two-level array
+    output. This will be used by other functions to store information about
+    SWAP, PauliX gates etc.
+    """
+    num_two_level_gates = len(two_level_output)
+
+    # create a reversed list of keys based on total number of two-level gates
+    # ranging from 1 to n where n is the total number of two-level arrays
+    gate_keys = list(range(1, num_two_level_gates + 1))[::-1]
+    gate_info_dict = dict.fromkeys(gate_keys)
+    return gate_info_dict
+
+
+def _partial_gray_code(num_qubits, two_level_output):
+    """Returns a dictionary of partial gray code sequence for each two-level
+    array.
+
+    The input is when output from decomposition array output is non-expanded."""
+
+    # create empty dict
+    gate_key_dict = _create_dict_for_two_level_arrays(two_level_output)
+
+    # create a list of non-trivial indices in two level array output
+    two_level_indices = []
+    for i in range(len(two_level_output)):
+        two_level_indices.append(two_level_output[i][0])
+
+    # gray code sequence output as indices of binary sequence and strings
+    # respectively
+    gray_code_index = _gray_code_sequence(num_qubits, "index_values")
+    gray_code_string = _gray_code_sequence(num_qubits)
+
+    # get the partial gray code sequence
+    for i in range(len(two_level_indices)):
+        partial_gray_code = []
+        ind1 = two_level_indices[i][0]
+        ind2 = two_level_indices[i][1]
+
+        ind1_pos_in_gray_code = gray_code_index.index(ind1)
+        ind2_pos_in_gray_code = gray_code_index.index(ind2)
+
+        if ind1_pos_in_gray_code > ind2_pos_in_gray_code:
+            partial_gray_code = [ind2_pos_in_gray_code, ind1_pos_in_gray_code]
+        else:
+            partial_gray_code = [ind1_pos_in_gray_code, ind2_pos_in_gray_code]
+
+        gate_key_dict[len(two_level_indices) - i] = gray_code_string[
+            partial_gray_code[0] : partial_gray_code[1] + 1
+        ]
+
+    return gate_key_dict
+
```

Comment:
Add a function to make all control values consistent before the partial gray code gets split into n-bit toffoli and the gate needing decomposition. 

If the function is used before ` _split_partial_gray_code` then don't have to worry about searching for controls between the two split dictionaries.